algorithms/linfa-linear/src/isotonic.rs

```diff
@@ -0,0 +1,304 @@
+/// let pred = model.predict(&dataset);
+/// let r2 = pred.r2(&dataset).unwrap();
+/// println!("r2 from prediction: {}", r2);
+/// ```
+pub struct IsotonicRegression {}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+/// A fitted isotonic regression model which can be used for making predictions.
+pub struct FittedIsotonicRegression<F> {
+    regressor: Array1<F>,
+    response: Array1<F>,
+}
+
+impl Default for IsotonicRegression {
+    fn default() -> Self {
+        IsotonicRegression::new()
+    }
+}
+
+/// Configure and fit a isotonic regression model
+impl IsotonicRegression {
+    /// Create a default isotonic regression model.
+    pub fn new() -> IsotonicRegression {
+        IsotonicRegression {}
+    }
+}
+
+impl<F: Float, D: Data<Elem = F>, T: AsSingleTargets<Elem = F>>
+    Fit<ArrayBase<D, Ix2>, T, LinearError<F>> for IsotonicRegression
+{
+    type Object = FittedIsotonicRegression<F>;
+
+    /// Fit an isotonic regression model given a feature matrix `X` and a target
+    /// variable `y`.
+    ///
+    /// The feature matrix `X` must have shape `(n_samples, 1)`
+    ///
+    /// The target variable `y` must have shape `(n_samples)`
+    ///
+    /// Returns a `FittedIsotonicRegression` object which contains the fitted
+    /// parameters and can be used to `predict` values of the target variable
+    /// for new feature values.
+    fn fit(&self, dataset: &DatasetBase<ArrayBase<D, Ix2>, T>) -> Result<Self::Object, F> {
+        let X = dataset.records();
+        let (n_samples, dim) = X.dim();
+        let y = dataset.as_single_targets();
+
+        // Check the input dimension
+        assert_eq!(dim, 1, "The input dimension must be 1.");
+
+        // Check that our inputs have compatible shapes
+        assert_eq!(y.dim(), n_samples);
+
+        // use correlation for determining relationship between x & y
+        let x = X.slice(s![.., 0]);
+        let rho = DatasetBase::from(stack![Axis(1), x, y]).pearson_correlation();
+        let incresing = rho.get_coeffs()[0] >= F::zero();
+
+        // sort data
+        let indices = argsort_by(&x, |a, b| a.partial_cmp(b).unwrap_or(Ordering::Greater));
+
+        // PVA algorithm
+        let mut J: Vec<(F, F, Vec<usize>)> = indices
+            .iter()
+            .map(|&i| (y[i], F::cast(dataset.weight_for(i)), vec![i]))
+            .collect();
+        if !incresing {
+            J.reverse();
+        };
+
+        let mut i: usize = 0;
+        while i < (J.len() - 1) {
+            let B_zero = &J[i];
+            let B_plus = &J[i + 1];
+            if B_zero.0 <= B_plus.0 {
+                i += 1;
+            } else {
+                let w0 = B_zero.1 + B_plus.1;
+                let v0 = (B_zero.0 * B_zero.1 + B_plus.0 * B_plus.1) / w0;
+                let mut i0 = B_zero.2.clone();
+                i0.extend(&(B_plus.2));
+                J[i] = (v0, w0, i0);
+                J.remove(i + 1);
```

Comment:
Vector removal is O(n), which is inefficient inside loops. Can you change `J` to a different data structure so removals are less expensive? `Hashmap` or `Vec<Option<...>>` are good options, but you'll need to change the way you index into `J`.